Add predictor how-to article (#8697)
MicrosoftDocs/PowerShell-Docs: reference/docs-conceptual/dev-cross-plat/create-cmdline-predictor.md

---
description: This article describes how to create a command-line predictor to help with command completion in PowerShell.
ms.date: 03/28/2022
title: How to create a command-line predictor
---
# How to create a command-line predictor

PSReadLine 2.1.0 introduced the concept of a smart command-line predictor by implementing the
Predictive IntelliSense feature. PSReadLine 2.2.2 expanded on that feature by adding a plugin model
that allows you create your own command-line predictors.

Predictive IntelliSense enhances tab completion by providing suggestions, on the command line, as
you type. The prediction suggestion appears as colored text following your cursor. This enables you

> to discover, edit, and execute full commands based on matching predictions from the user's history
> or additional domain-specific plugins.

Reviewer: In this context, `you` is referring to the person using the Predictive IntelliSense, not (necessarily) the person _developing a predictor_, right?
```suggestion
to discover, edit, and execute full commands based on matching predictions from your command
history or additional domain-specific plugins.
```

## System requirements

To create and use a plugin predictor, you must be using the following versions of software:

- PowerShell 7.2 (or higher) - provides the APIs necessary for creating a command predictor
- PSReadLine 2.2.2 (or higher) - allows you to configure PSReadLine to use the plugin

## Overview of a predictor

A predictor is a PowerShell binary module. The module must implement the
`System.Management.Automation.Subsystem.Prediction.ICommandPredictor` interface. This interface
declares the methods used to query for prediction results and provide feedback.

A predictor module must register a `CommandPredictor` subsystem with PowerShell's `SubsystemManager`
when loaded and unregister itself when unloaded.

The following diagram shows the architecture of a predictor in PowerShell.

![Architecture](media/create-cmdline-predictor/predictor-architecture.png)

## Creating the code

To create a predictor, you must have the .NET 6 SDK installed for your platform. For more
information on the SDK, see [Download .NET 6.0](https://dotnet.microsoft.com/download/dotnet/6.0).

> Create a new PowerShell module project using the following commands:

Reviewer: I'm not sure on "using the following steps" vs "with the following steps" here. Maybe "by following these steps?

```suggestion
Create a new PowerShell module project by following these steps:
```

Use the `dotnet` command-line tool to create a starter classlib project.

```powershell
dotnet new classlib --name SamplePredictor
```

Edit the `SamplePredictor.csproj` to contain the following information:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.PowerShell.SDK" Version="7.2.0" />
  </ItemGroup>

</Project>
```

The following example code returns the string "HELLO WORLD" for the prediction result for all user
input. Since the sample predictor doesn't process any feedback, the code does not implement the
feedback methods from the interface. Change the prediction and feedback code to meet the needs of

> your predictor.

Reviewer: I think this should move to after the next paragraph and possibly be in a note alert. Later in the document, the example assumes the developer has left the sample code untouched.

        /// <param name="commandLine">The last accepted command line.</param>
        /// <param name="success">Shows whether the execution was successful.</param>
        public void OnCommandLineExecuted(PredictionClient client, string commandLine, bool success) { }

        #endregion;
    }

    /// <summary>
    /// Register the predictor on module loading and unregister it on module un-loading.
    /// </summary>
    public class Init : IModuleAssemblyInitializer, IModuleAssemblyCleanup
    {
        private const string Identifier = "843b51d0-55c8-4c1a-8116-f0728d419306";

        /// <summary>
        /// Gets called when assembly is loaded.
        /// </summary>
        public void OnImport()
        {
            var predictor = new SamplePredictor(Identifier);
            SubsystemManager.RegisterSubsystem(SubsystemKind.CommandPredictor, predictor);
        }

        /// <summary>
        /// Gets called when the binary module is unloaded.
        /// </summary>
        public void OnRemove(PSModuleInfo psModuleInfo)
        {
            SubsystemManager.UnregisterSubsystem(SubsystemKind.CommandPredictor, new Guid(Identifier));
        }
    }
}
```

> Once you created the project with the sample code, run `dotnet build` to produce the assembly, for
> example, `bin/Debug/net6.0/SamplePredictor.dll`.

Reviewer: ```suggestion
Once you have created the project with the sample code, run `dotnet build` to produce the assembly,
for example, `bin/Debug/net6.0/SamplePredictor.dll`.
```
